File: cpp/src/generation/tpms/tpms.cpp
#include "tpms.h"


// Generation of triply periodic minimal surface (TPMS)

//if you want the result to be continuous (non-thresholded)
bool puma::generateTPMS(puma::Workspace *tpmsWS, TPMSInput input) {

    Generate_TPMS generator(tpmsWS,input);
    return generator.generate();
}


Generate_TPMS::Generate_TPMS(puma::Workspace *tpmsWS, TPMSInput input) {
    this->tpmsWS = tpmsWS;
    this->input = input;
}


bool Generate_TPMS::generate() {

    if( !logInput() ) {
        std::cout << "Failed to log Inputs" << std::endl;
        return false;
    }

    std::string errorMessage;
    if( !errorCheck(&errorMessage) ) { //Error Check Failed
        std::cout << "Error in TPMS Generation: " << errorMessage << std::endl;
        return false;
    }

    bool success = TPMSGenerationHelper();
    if(!success){
        return false;
    }

    if( !logOutput() ) {
        std::cout << "Failed to log Outputs" << std::endl;
        return false;
    }

    return true;
}


bool Generate_TPMS::TPMSGenerationHelper() {

    double wMin = input.wMin;
    double wMax = input.wMax;
    double qMin = input.qMin;
    double qMax = input.qMax;

    puma::Matrix<double> mat(input.xSize,input.ySize,input.zSize);
    tpmsWS->matrix.resize(input.xSize,input.ySize,input.zSize);

#pragma omp parallel for
    for(int i=0;i<input.xSize;i++) {
        for(int j=0;j<input.ySize;j++) {
            for(int k=0;k<input.zSize;k++) {


                double q = qMin + ((double)k/(double)(input.zSize-1))*(qMax-qMin);
                double w = wMin + ((double)k/(double)(input.zSize-1))*(wMax-wMin);

                if(input.equation == 0) {
                    mat(i, j, k) =
                            std::sin(w * i) * std::sin(w * j) * std::sin(w * k) +
                            std::sin(w * i) * std::cos(w * j) * std::cos(w * k) +
                            std::cos(w * i) * std::sin(w * j) * std::cos(w * k) +
                            std::cos(w * i) * std::cos(w * j) * std::sin(w * k) + q;
                }
                else if(input.equation == 1) {
                    mat(i, j, k) =
                            std::cos(w * i) * std::sin(w * j) +
                            std::cos(w * j) * std::sin(w * k) +
                            std::cos(w * k) * std::cos(w * i) + q;
                }
                else if(input.equation == 2) {
                    mat(i, j, k) =
                             std::cos(w * i) + std::cos(w * j) + std::cos(w * k) + q;
                }
            }
        }
    }


    std::cout << mat.max() << " " << mat.min() << std::endl;

    for(long i=0;i<mat.size();i++) {
        if(mat(i) < 0.8) mat(i)=0.8;
        if(mat(i) > 1.2) mat(i)=1.2;
    }

     std::cout << mat.max() << " " << mat.min() << std::endl;

    mat.normalize(0,255);


#pragma omp parallel for
    for(int i=0;i<input.xSize;i++) {
        for(int j=0;j<input.ySize;j++) {
            for(int k=0;k<input.zSize;k++) {
                 tpmsWS->matrix(i,j,k)=(short)mat(i,j,k);
            }
        }
    }


    return true;
}


bool Generate_TPMS::errorCheck(std::string *errorMessage) {

    if(input.xSize <= 0) {
        (*errorMessage) = "xSize invalid. Must be >0";
        return false;
    }
    if(input.ySize <= 0) {
        (*errorMessage) = "ySize invalid. Must be >0";
        return false;
    }
    if(input.zSize <= 0) {
        (*errorMessage) = "zSize invalid. Must be >0";
        return false;
    }
    if(input.wMin <= 0) {
        (*errorMessage) = "wMin invalid. Must be >0";
        return false;
    }
    if(input.wMax < input.wMin) {
        (*errorMessage) = "wMin invalid. Must be >=0";
        return false;
    }
    if(input.qMin <= 0) {
        (*errorMessage) = "qMin invalid. Must be >0";
        return false;
    }
    if(input.qMax < input.qMin) {
        (*errorMessage) = "qMax invalid. Must be >=0";
        return false;
    }

    if(input.equation != 0 && input.equation != 1 && input.equation != 2) {
        (*errorMessage) = "Invalid input equation. Must be 0, 1, or 2";
        return false;
    }

    return true;
}


bool Generate_TPMS::logInput() {
    puma::Logger *log;
    log = tpmsWS->log;

    (*log).appendLogSection("Generate TPMS Structure");
    (*log).appendLogItem("Time: ");
    (*log).appendLogItem(puma::Logger::getTime());
    (*log).newLine();
    (*log).appendLogLine(" -- Inputs:");
    (*log).appendLogItem("Domain Size: ");
    (*log).appendLogItem(input.xSize);
    (*log).appendLogItem(", ");
    (*log).appendLogItem(input.ySize);
    (*log).appendLogItem(", ");
    (*log).appendLogItem(input.zSize);
    (*log).newLine();
    (*log).appendLogItem("wMin: ");
    (*log).appendLogItem(input.wMin);
    (*log).newLine();
    (*log).appendLogItem("wMax: ");
    (*log).appendLogItem(input.wMax);
    (*log).newLine();
    (*log).appendLogItem("qMin: ");
    (*log).appendLogItem(input.qMin);
    (*log).newLine();
    (*log).appendLogItem("qMax: ");
    (*log).appendLogItem(input.qMax);
    (*log).newLine();
    (*log).appendLogItem("Equation ID: ");
    (*log).appendLogItem(input.equation);
    (*log).newLine();

    (*log).writeLog();

    return true;

}


bool Generate_TPMS::logOutput() {
    puma::Logger *log;
    log = tpmsWS->log;

    (*log).appendLogItem("Successful TPMS Generation");
    (*log).newLine();
    (*log).appendLogItem("Time: ");
    (*log).appendLogItem(puma::Logger::getTime());
    (*log).newLine();

    (*log).writeLog();

    return true;
}

File: python/pumapy/generation/tpms.py
from pumapy.utilities.workspace import Workspace
from pumapy.utilities.timer import Timer
from pumapy.generation.tpms_utils import generate
import numpy as np


def generate_tpms(shape, w, q, equation=0, segmented=True):
    """ Generation of triply periodic minimal surface material

        :param shape: shape of 3D domain (x,y,z)
        :type shape: (int, int, int)
        :param w: w parameter for tpms
        :type w: float or (float, float)
        :param q: q parameter for tpms (float or tuple)
        :type q: float or (float, float)
        :param equation: equation 0, 1, or 2 for tpms
        :type equation: int
        :param segmented: return a domain that is already segmented (i.e. 1 for solid and 0 for void) or
            with grayscales 0-255 with threshold=128 for the solid
        :type segmented: bool
        :return: TPMS domain with grayscales from 0-255 with threshold=128 for the normal TPMS geometry
        :rtype: Workspace

        :Example:
        >>> import pumapy as puma
        >>> shape = (200, 200, 200)  # size of the domain, in voxels.
        >>> w = 0.08  # value of w in the equations above
        >>> q = 0.2  # value of q in the equations above
        >>> ws_eq0 = puma.generate_tpms(shape, w, q, equation=0, segmented=False)
        >>> ws_eq1 = puma.generate_tpms(shape, w, q, equation=1, segmented=False)
        >>> ws_eq2 = puma.generate_tpms(shape, w, q, equation=2, segmented=False)
        >>> puma.render_contour(ws_eq0, cutoff=(128, 255)) #visualize the workspace
        >>> puma.render_contour(ws_eq1, cutoff=(128, 255))
        >>> puma.render_contour(ws_eq2, cutoff=(128, 255))
    """

    if isinstance(w, tuple):
        _w = w
    elif isinstance(w, float):
        _w = (w, w)
    else:
        raise Exception("Invalid w, must be float or tuple")

    if isinstance(q, tuple):
        _q = q
    elif isinstance(q, float):
        _q = (q, q)
    else:
        raise Exception("Invalid w, must be float or tuple")

    generator = GeneratorTPMS(shape, _w, _q, equation)

    generator.error_check()

    generator.log_input()
    ws = generator.generate()
    generator.log_output()

    if segmented:
        ws.binarize(128)
    return ws


class GeneratorTPMS:
    def __init__(self, size, w, q, equation):
        self._workspace = Workspace()
        self._size = size
        self._wmin = w[0]
        self._wmax = w[1]
        self._qmin = q[0]
        self._qmax = q[1]
        self._equation = equation

    def generate(self):
        t = Timer()
        self._workspace.resize_new_matrix(self._size)
        self._workspace.matrix = generate(self._workspace.matrix.astype(float), self._size[0], self._size[1],
                                          self._size[2], self._wmin, self._wmax, self._qmin, self._qmax, self._equation)
        self._workspace.matrix = (255. * (self._workspace.matrix - np.min(self._workspace.matrix)) /
                                  np.ptp(self._workspace.matrix)).astype(np.uint16)
        print("Generated in: " + str(t.elapsed()) + " seconds")
        return self._workspace

    def log_input(self):
        self._workspace.log.log_section("Generating TPMS")
        self._workspace.log.log_line("Domain Size: " + str(self._size))
        self._workspace.log.log_line("w: " + str(self._wmin) + ", " + str(self._wmax))
        self._workspace.log.log_line("w: " + str(self._qmin) + ", " + str(self._qmax))
        self._workspace.log.log_line("Equation: " + str(self._equation))
        self._workspace.log.write_log()

    def log_output(self):
        self._workspace.log.log_section("Finished TPMS Generation")
        self._workspace.log.write_log()

    def error_check(self):
        if not isinstance(self._workspace, Workspace):
            raise Exception("Error, workspace must be passed")
        if not (isinstance(self._size, tuple) and len(self._size) == 3):
            raise Exception("Error, invalid size, must be tuple of size 3")
        if self._size[0] <= 0:
            raise Exception("Error, invalid size, must be >= 0")
        if self._size[1] <= 0:
            raise Exception("Error, invalid size, must be >= 0")
        if self._size[2] <= 0:
            raise Exception("Error, invalid size, must be >= 0")
        if self._wmin <= 0:
            raise Exception("Error, invalid w, must be >= 0")
        if self._wmax <= 0:
            raise Exception("Error, invalid w, must be >= 0")
        if self._wmax < self._wmin:
            raise Exception("Error, invalid w, w[1] cannot be < w[0]")
        if self._qmin <= 0:
            raise Exception("Error, invalid q, must be >= 0")
        if self._qmax <= 0:
            raise Exception("Error, invalid q, must be >= 0")
        if self._qmax < self._wmin:
            raise Exception("Error, invalid q, q[1] cannot be < q[0]")
        if self._equation != 0 and self._equation != 1 and self._equation != 2:
            raise Exception("Error, invalid equation number. Must be 0, 1, or 2")
